R/raincloudPlots.R

```diff
@@ -0,0 +1,908 @@
+  }
+
+  if (options$mean && (meanInterval || options$meanIntervalCustom)) {
+    tableInProgress$addColumnInfo(
+      name = "upperBound", title = "Upper Interval Limit", type = "number", format = "dp:2"
+    )
+  }
+
+  tableInProgress$showSpecifiedColumnsOnly <- TRUE  # Only show columns that were added
+
+  return(tableInProgress)
+}  # End .rainAddTableColumns()
+
+
+
+# .rainFillTable() ----
+.rainFillTable <- function(jaspResults, dataInfo, inputVariable, options, inputTable) {
+
+  # Extract statistics from corresponding plot
+  plotContainer <- jaspResults[["containerRaincloudPlots"]]
+  targetJaspPlot <- plotContainer[[inputVariable]]
+  targetPlotObject <- targetJaspPlot[["plotObject"]]
+
+  # Extract observed factor combinations and remove duplicate levels of primaryFactor for table
+  combinations <- .rainInfoFactorCombinations(dataInfo$dataset, targetPlotObject, extractColor = FALSE)$uniqueCombis
+  primaryLevels <- as.character(combinations$primaryFactor)
+  previousLevel <- primaryLevels[1]
+  tidyLevels <- c(previousLevel)
+  if (length(primaryLevels) > 1) {
+    for (i in 2:length(primaryLevels)) {
+      if (primaryLevels[i] == previousLevel) {
+        tidyLevels <- c(tidyLevels, "")
+      } else {
+        tidyLevels <- c(tidyLevels, primaryLevels[i])
+        previousLevel <- primaryLevels[i]
+      }
+    }
+  }
+  combinations$primaryFactor <- tidyLevels
+
+  boxDataIndex <- if (options$observationId == "") 2 else 3
+  boxData <- ggplot2::ggplot_build(targetPlotObject)$data[[boxDataIndex]]
+
+  tableStatistics <- cbind(combinations, boxData)
+
+  if (options$mean) {
+    meanDataIndex <- if (options$observationId == "") 6 else 7
+    tableStatistics$mean <- ggplot2::ggplot_build(targetPlotObject)$data[[meanDataIndex]]$y_orig
+  }
+  if (options$meanInterval || options$meanIntervalCustom) {
+    tableStatistics$lowerBound <- dataInfo$intervalBounds[[inputVariable]]$lowerBound
+    tableStatistics$sd         <- dataInfo$intervalBounds[[inputVariable]]$sd
+    tableStatistics$upperBound <- dataInfo$intervalBounds[[inputVariable]]$upperBound
+  }
+
+  # Add statistics to table
+  inputTable$setData(tableStatistics)
+
+  # Footnote
+  sampleSize <- gettextf("<i>N</i><sub>Total</sub> = %s.", dataInfo$sampleSize)
+  exclusions <- if (dataInfo$numberOfExclusions > 0) {
+    gettextf(" This excludes %s observations due to missing data.", dataInfo$numberOfExclusions)
+  } else {
+    NULL
+  }
+  meanInterval <- if (options$meanIntervalCustom) {
+    gettextf(" Interval around mean is custom.")
+  } else if (options$observationId == "" && options$meanInterval && options$meanIntervalOption == "ci" && options$meanCiAssumption) {
+    paste0(
+      " Interval around mean represents ",
+      options$meanCiWidth * 100,
+      "% confidence interval; confidence intervals were computed independently for each group.")
+  } else {
+    NULL
+  }
+
+  footnote <- paste0(sampleSize, exclusions, meanInterval)
```

Comment:
Often, it is better to write one series of if statements _with_ code duplication. For example:

```r
footnote <- NULL
if (options$meanIntervalCustom) {
  if (dataInfo$numberOfExclusions == 0 && 
	  !(options$meanIntervalCustom || !(options$observationId == "" && options$meanInterval && options$meanIntervalOption == "ci" && options$meanCiAssumption)))
    gettextf("<i>N</i><sub>Total</sub> = %s.", dataInfo$sampleSize)
  else if (dataInfo$numberOfExclusions == 0) {
    if (options$meanIntervalCustom) {
      gettextf("<i>N</i><sub>Total</sub> = %s. Interval around mean is custom.", dataInfo$sampleSize)
    } else {
      # etc.
    }
  }
```
While the code becomes objectively worse, the individual strings become easier to translate. Now, in this case, I think the sentences are distinct enough that it doesn't really matter so whether you want to do this is up to you.

Reply:
Because it is still possible to translate the strings, I would like to keep the code tidier.
I made minor changes so that the footnote works even if there are no exclusions or meanInterval,
while translators do not have to wrangle with random spaces at the beginning of a string.